File: backend/services/webSearchService.js
/**
 * Web Search Service using Serper API
 * Provides intelligent web search capabilities for the wealth advisor
 */

const axios = require('axios');

class WebSearchService {
  constructor() {
    this.apiKey = process.env.SERPER_API_KEY;
    this.baseUrl = 'https://google.serper.dev/search';
    this.defaultLocation = 'India';
    this.defaultGl = 'in'; // India
    
    // Log configuration status (without exposing the actual key)
    console.log('Web Search Service initialized');
    console.log('Serper API configured:', !!this.apiKey);
  }

  /**
   * Perform web search using Serper API
   * @param {string} query - Search query
   * @param {Object} options - Additional search options
   * @returns {Promise<Object>} Search results
   */
  async search(query, options = {}) {
    if (!this.apiKey) {
      console.warn('Serper API key not configured');
      return { error: 'Web search not available' };
    }

    try {
      const searchParams = {
        q: query,
        location: options.location || this.defaultLocation,
        gl: options.gl || this.defaultGl,
        num: options.num || 5, // Limit results for faster response
        ...options
      };

      const response = await axios.post(this.baseUrl, searchParams, {
        headers: {
          'X-API-KEY': this.apiKey,
          'Content-Type': 'application/json'
        },
        timeout: 10000 // 10 second timeout
      });

      return this.formatSearchResults(response.data);
    } catch (error) {
      console.error('Web search error:', error.message);
      return { error: 'Failed to perform web search' };
    }
  }

  /**
   * Format search results for better readability
   * @param {Object} rawResults - Raw search results from Serper
   * @returns {Object} Formatted results
   */
  formatSearchResults(rawResults) {
    if (!rawResults || rawResults.error) {
      return { error: rawResults?.error || 'No search results available' };
    }

    const results = {
      searchQuery: rawResults.searchParameters?.q || '',
      totalResults: rawResults.searchInformation?.totalResults || 0,
      searchTime: rawResults.searchInformation?.searchTime || 0,
      organic: [],
      answerBox: null,
      knowledgeGraph: null
    };

    // Format organic results
    if (rawResults.organic && Array.isArray(rawResults.organic)) {
      results.organic = rawResults.organic.slice(0, 5).map(result => ({
        title: result.title,
        link: result.link,
        snippet: result.snippet,
        date: result.date
      }));
    }

    // Include answer box if available
    if (rawResults.answerBox) {
      results.answerBox = {
        answer: rawResults.answerBox.answer,
        title: rawResults.answerBox.title,
        link: rawResults.answerBox.link
      };
    }

    // Include knowledge graph if available
    if (rawResults.knowledgeGraph) {
      results.knowledgeGraph = {
        title: rawResults.knowledgeGraph.title,
        type: rawResults.knowledgeGraph.type,
        description: rawResults.knowledgeGraph.description
      };
    }

    return results;
  }

  /**
   * Determine if a query needs web search based on content analysis
   * @param {string} query - User query
   * @param {Object} userContext - User's financial context
   * @returns {boolean} Whether web search is needed
   */
  shouldPerformWebSearch(query, userContext = {}) {
    const lowerQuery = query.toLowerCase();
    
    // Keywords that indicate need for current information
    const webSearchTriggers = [
      // Market data queries
      'current price', 'today price', 'latest price', 'stock price',
      'market today', 'nifty today', 'sensex today', 'market update',
      'current market', 'today market', 'market news',
      
      // Current events and news
      'news', 'latest news', 'recent news', 'today news',
      'what happened', 'recent developments', 'latest update',
      'major news', 'breaking news', 'headlines',
      
      // Holdings and portfolio specific
      'holdings news', 'my holdings', 'portfolio news',
      'stock news', 'investment news',
      
      // Economic indicators
      'interest rate', 'repo rate', 'inflation rate', 'gdp',
      'economic news', 'rbi news', 'budget news',
      
      // Specific company news
      'earnings', 'quarterly results', 'annual report',
      'company news', 'stock split', 'dividend',
      
      // Investment opportunities
      'new ipo', 'upcoming ipo', 'new schemes', 'new funds',
      'investment opportunities', 'market opportunities',
      
      // Regulatory changes
      'new rules', 'sebi news', 'tax changes', 'policy changes'
    ];

    // Check if query contains web search triggers
    const needsWebSearch = webSearchTriggers.some(trigger => 
      lowerQuery.includes(trigger)
    );

    // Additional logic: Check if query mentions specific stocks from user's portfolio
    if (!needsWebSearch && userContext.portfolio?.stocks) {
      const stockSymbols = userContext.portfolio.stocks.map(stock => 
        stock.symbol.toLowerCase()
      );
      
      const mentionsUserStock = stockSymbols.some(symbol => 
        lowerQuery.includes(symbol) || 
        lowerQuery.includes(symbol.replace('ltd', '').trim())
      );
      
      if (mentionsUserStock && (
        lowerQuery.includes('how is') || lowerQuery.includes('performance') ||
        lowerQuery.includes('doing') || lowerQuery.includes('price')
      )) {
        return true;
      }
    }

    return needsWebSearch;
  }

  /**
   * Generate search query based on user's question and context
   * @param {string} userQuery - Original user query
   * @param {Object} userContext - User's financial context
   * @returns {string} Optimized search query
   */
  generateSearchQuery(userQuery, userContext = {}) {
    const lowerQuery = userQuery.toLowerCase();
    
    // Holdings news queries
    if (lowerQuery.includes('holdings news') || 
        lowerQuery.includes('portfolio news') ||
        (lowerQuery.includes('news') && (lowerQuery.includes('holdings') || lowerQuery.includes('my')))) {
      
      if (userContext.portfolio?.stocks && userContext.portfolio.stocks.length > 0) {
        // Get user's stock symbols for targeted search
        const stockSymbols = userContext.portfolio.stocks.slice(0, 3).map(stock => stock.symbol);
        return `${stockSymbols.join(' ')} latest news stock market India today`;
      }
    }
    
    // If query is about specific stocks in user's portfolio
    if (userContext.portfolio?.stocks) {
      const stockSymbols = userContext.portfolio.stocks.map(stock => stock.symbol);
      const mentionedStock = stockSymbols.find(symbol => 
        lowerQuery.includes(symbol.toLowerCase())
      );
      
      if (mentionedStock) {
        if (lowerQuery.includes('price') || lowerQuery.includes('today')) {
          return `${mentionedStock} stock price today NSE BSE`;
        }
        if (lowerQuery.includes('news') || lowerQuery.includes('update')) {
          return `${mentionedStock} stock news latest updates`;
        }
      }
    }

    // Market queries
    if (lowerQuery.includes('nifty') || lowerQuery.includes('sensex')) {
      return `Nifty Sensex today market update India`;
    }

    // Interest rate queries
    if (lowerQuery.includes('interest rate') || lowerQuery.includes('repo rate')) {
      return `RBI repo rate interest rates India latest`;
    }

    // Default: return original query with India context
    return `${userQuery} India finance investment`;
  }

  /**
   * Summarize search results for AI consumption
   * @param {Object} searchResults - Formatted search results
   * @returns {string} Summary text for AI
   */
  summarizeSearchResults(searchResults) {
    if (searchResults.error) {
      return `Web search unavailable: ${searchResults.error}`;
    }

    let summary = `CURRENT WEB SEARCH RESULTS for "${searchResults.searchQuery}":\n\n`;

    // Add answer box if available
    if (searchResults.answerBox) {
      summary += `DIRECT ANSWER: ${searchResults.answerBox.answer}\nSource: ${searchResults.answerBox.title} (${searchResults.answerBox.link})\n\n`;
    }

    // Add knowledge graph if available
    if (searchResults.knowledgeGraph) {
      summary += `KEY INFORMATION: ${searchResults.knowledgeGraph.title} - ${searchResults.knowledgeGraph.description}\n\n`;
    }

    // Add top organic results with detailed information
    if (searchResults.organic && searchResults.organic.length > 0) {
      summary += 'LATEST INFORMATION FROM WEB SOURCES:\n';
      searchResults.organic.slice(0, 4).forEach((result, index) => {
        summary += `${index + 1}. TITLE: ${result.title}\n`;
        summary += `   CONTENT: ${result.snippet}\n`;
        summary += `   SOURCE: ${result.link}\n`;
        if (result.date) summary += `   PUBLISHED: ${result.date}\n`;
        summary += `   ---\n`;
      });
    }

    summary += `\nIMPORTANT INSTRUCTIONS:\n`;
    summary += `- Use this current, real-time information in your response\n`;
    summary += `- Always cite sources using [Source: Title](link) format\n`;
    summary += `- Mention specific data points from the search results\n`;
    summary += `- Include "Sources:" section at the end with numbered references\n`;

    return summary;
  }

  /**
   * Extract source links from search results for citation
   * @param {Object} searchResults - Formatted search results
   * @returns {Array} Array of source objects with title and link
   */
  extractSources(searchResults) {
    if (searchResults.error || !searchResults.organic) {
      return [];
    }

    const sources = [];
    
    // Add answer box source if available
    if (searchResults.answerBox && searchResults.answerBox.link) {
      sources.push({
        title: searchResults.answerBox.title,
        link: searchResults.answerBox.link
      });
    }

    // Add organic result sources
    searchResults.organic.slice(0, 4).forEach(result => {
      sources.push({
        title: result.title,
        link: result.link,
        date: result.date
      });
    });

    return sources;
  }
}

module.exports = new WebSearchService();

File: backend_python/services/web_search_service.py
"""
Web search service using Serper API for real-time market data and news
"""

import logging
from typing import Dict, Any, List, Optional
import httpx
import asyncio
from datetime import datetime

from config.settings import get_settings
from models.chat import WebSearchResult, WebSearchResponse

logger = logging.getLogger(__name__)
settings = get_settings()


class WebSearchService:
    def __init__(self):
        self.api_key = settings.serper_api_key
        self.base_url = "https://google.serper.dev/search"
        self.default_location = "India"
        self.default_gl = "in"  # India
        
        logger.info("Web Search Service initialized")
        logger.info(f"Serper API configured: {bool(self.api_key)}")
    
    async def search(self, query: str, options: Optional[Dict[str, Any]] = None) -> Dict[str, Any]:
        """Perform web search using Serper API"""
        
        if not self.api_key:
            logger.warning("Serper API key not configured")
            return {"error": "Web search not available"}
        
        if not options:
            options = {}
        
        # Prepare search parameters
        search_params = {
            "q": query,
            "gl": options.get("gl", self.default_gl),
            "hl": "en",
            "num": options.get("num", 5),
            "type": "search"
        }
        
        # Add location if specified
        if "location" in options:
            search_params["location"] = options["location"]
        
        headers = {
            "X-API-KEY": self.api_key,
            "Content-Type": "application/json"
        }
        
        try:
            async with httpx.AsyncClient() as client:
                response = await client.post(
                    self.base_url,
                    json=search_params,
                    headers=headers,
                    timeout=30.0
                )
                
                if response.status_code == 200:
                    results = response.json()
                    logger.info(f"Web search successful for query: {query}")
                    return results
                else:
                    logger.error(f"Web search failed with status {response.status_code}: {response.text}")
                    return {"error": f"Search request failed: {response.status_code}"}
        
        except httpx.TimeoutException:
            logger.error("Web search request timed out")
            return {"error": "Search request timed out"}
        except Exception as e:
            logger.error(f"Web search error: {str(e)}")
            return {"error": f"Search service error: {str(e)}"}
    
    def should_perform_web_search(self, user_message: str, user_context: Optional[Dict[str, Any]] = None) -> bool:
        """Determine if web search should be performed for the given query"""
        
        # Keywords that typically require current information
        search_keywords = [
            "current", "today", "latest", "recent", "now", "update",
            "market", "price", "stock", "nifty", "sensex", "news",
            "economy", "rbi", "inflation", "budget", "policy",
            "performance", "trend", "analysis", "forecast"
        ]
        
        # Time-based keywords
        time_keywords = [
            "today", "this week", "this month", "this year",
            "recent", "latest", "current", "now"
        ]
        
        # Financial market keywords
        market_keywords = [
            "market", "stock", "share", "equity", "mutual fund",
            "nifty", "sensex", "bse", "nse", "trading",
            "price", "valuation", "analysis"
        ]
        
        message_lower = user_message.lower()
        
        # Check for search-triggering keywords
        has_search_keyword = any(keyword in message_lower for keyword in search_keywords)
        has_time_keyword = any(keyword in message_lower for keyword in time_keywords)
        has_market_keyword = any(keyword in message_lower for keyword in market_keywords)
        
        # Always search for time-sensitive market queries
        if has_time_keyword and has_market_keyword:
            return True
        
        # Search for general market queries
        if has_market_keyword and has_search_keyword:
            return True
        
        # Search for specific stock queries
        if any(symbol in message_lower for symbol in ['infy', 'tcs', 'hdfc', 'reliance', 'wipro']):
            return True
        
        # Check if user has specific stocks in portfolio and query is related
        if user_context and user_context.get('portfolio', {}).get('stocks'):
            user_stocks = [stock.get('symbol', '').lower() for stock in user_context['portfolio']['stocks']]
            if any(stock in message_lower for stock in user_stocks):
                return True
        
        return False
    
    def generate_search_query(self, user_message: str, user_context: Optional[Dict[str, Any]] = None) -> str:
        """Generate optimized search query for the user message"""
        
        message_lower = user_message.lower()
        
        # Market overview queries
        if any(keyword in message_lower for keyword in ["market", "overview", "update", "today"]):
            return "India stock market today Nifty Sensex latest news"
        
        # Specific stock queries
        stock_symbols = ['infy', 'tcs', 'hdfc', 'reliance', 'wipro', 'icici']
        for symbol in stock_symbols:
            if symbol in message_lower:
                return f"{symbol} stock price today India market news"
        
        # Economic policy queries
        if any(keyword in message_lower for keyword in ["rbi", "policy", "inflation", "interest"]):
            return "RBI monetary policy India interest rates inflation latest"
        
        # Mutual fund queries
        if "mutual fund" in message_lower or "mf" in message_lower:
            return "India mutual funds performance today market analysis"
        
        # Investment advice queries
        if any(keyword in message_lower for keyword in ["invest", "buy", "sell", "recommendation"]):
            return "India stock market investment advice today analysis"
        
        # Default: enhance the query with Indian market context
        enhanced_query = f"{user_message} India stock market latest"
        
        # Add user's stock context if available
        if user_context and user_context.get('portfolio', {}).get('stocks'):
            stocks = user_context['portfolio']['stocks']
            if isinstance(stocks, list) and stocks:
                top_stocks = stocks[:3]  # Top 3 holdings
                stock_symbols = [stock.get('symbol', '') for stock in top_stocks if isinstance(stock, dict)]
                if stock_symbols:
                    enhanced_query += f" {' '.join(stock_symbols)}"
        
        return enhanced_query
    
    def summarize_search_results(self, search_results: Dict[str, Any]) -> str:
        """Summarize search results for AI context"""
        
        if "error" in search_results:
            return f"Web search error: {search_results['error']}"
        
        summary = "\n=== CURRENT WEB SEARCH RESULTS ===\n"
        
        # Add search metadata
        if "searchParameters" in search_results:
            query = search_results["searchParameters"].get("q", "")
            summary += f"Search Query: {query}\n"
            summary += f"Search Time: {datetime.now().strftime('%Y-%m-%d %H:%M:%S IST')}\n\n"
        
        # Add organic results
        if "organic" in search_results:
            summary += "=== TOP SEARCH RESULTS ===\n"
            for i, result in enumerate(search_results["organic"][:5], 1):
                title = result.get("title", "")
                snippet = result.get("snippet", "")
                link = result.get("link", "")
                date = result.get("date", "")
                
                summary += f"{i}. **{title}**\n"
                if date:
                    summary += f"   Date: {date}\n"
                summary += f"   {snippet}\n"
                summary += f"   Source: {link}\n\n"
        
        # Add knowledge graph if available
        if "knowledgeGraph" in search_results:
            kg = search_results["knowledgeGraph"]
            summary += "=== KNOWLEDGE GRAPH INFO ===\n"
            if "title" in kg:
                summary += f"Title: {kg['title']}\n"
            if "description" in kg:
                summary += f"Description: {kg['description']}\n"
            if "attributes" in kg:
                for key, value in kg["attributes"].items():
                    summary += f"{key}: {value}\n"
            summary += "\n"
        
        # Add news results if available
        if "news" in search_results:
            summary += "=== LATEST NEWS ===\n"
            for i, news in enumerate(search_results["news"][:3], 1):
                title = news.get("title", "")
                snippet = news.get("snippet", "")
                date = news.get("date", "")
                source = news.get("source", "")
                
                summary += f"{i}. **{title}**\n"
                if source:
                    summary += f"   Source: {source}\n"
                if date:
                    summary += f"   Date: {date}\n"
                summary += f"   {snippet}\n\n"
        
        summary += "=== END WEB SEARCH RESULTS ===\n"
        
        return summary
    
    def extract_sources(self, search_results: Dict[str, Any]) -> List[Dict[str, str]]:
        """Extract source citations from search results"""
        
        sources = []
        
        if "organic" in search_results:
            for result in search_results["organic"][:5]:
                source = {
                    "title": result.get("title", ""),
                    "link": result.get("link", ""),
                    "date": result.get("date", "")
                }
                sources.append(source)
        
        if "news" in search_results:
            for news in search_results["news"][:3]:
                source = {
                    "title": news.get("title", ""),
                    "link": news.get("link", ""),
                    "date": news.get("date", ""),
                    "source": news.get("source", "")
                }
                sources.append(source)
        
        return sources
    
    async def search_market_data(self, query: str) -> Dict[str, Any]:
        """Specialized search for market data"""
        market_query = f"{query} India stock market Nifty Sensex BSE NSE latest"
        return await self.search(market_query, {"location": "India", "num": 10})
    
    async def search_company_news(self, company_symbol: str) -> Dict[str, Any]:
        """Search for specific company news"""
        query = f"{company_symbol} stock price news India market analysis latest"
        return await self.search(query, {"location": "India", "num": 5})
    
    async def search_economic_news(self) -> Dict[str, Any]:
        """Search for general economic news"""
        query = "India economy RBI policy inflation interest rates stock market latest news"
        return await self.search(query, {"location": "India", "num": 8})